build/utf8/valid_asm.go

```diff
@@ -0,0 +1,547 @@
+
+	// The default lowest and highest continuation byte.
+	locb = 0b10000000 // 128 0x80
+	hicb = 0b10111111 // 191 0xBF
+
+	// These names of these constants are chosen to give nice alignment in the
+	// table below. The first nibble is an index into acceptRanges or F for
+	// special one-byte cases. The second nibble is the Rune length or the
+	// Status for the special one-byte case.
+	xx = 0xF1 // invalid: size 1
+	as = 0xF0 // ASCII: size 1
+	s1 = 0x02 // accept 0, size 2
+	s2 = 0x13 // accept 1, size 3
+	s3 = 0x03 // accept 0, size 3
+	s4 = 0x23 // accept 2, size 3
+	s5 = 0x34 // accept 3, size 4
+	s6 = 0x04 // accept 0, size 4
+	s7 = 0x44 // accept 4, size 4
+)
+
+// TODO: find the address of this from unicode/utf8.
+// first is information about the first byte in a UTF-8 sequence.
+var firstData = [256]uint8{
+	//   1   2   3   4   5   6   7   8   9   A   B   C   D   E   F
+	as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, // 0x00-0x0F
+	as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, // 0x10-0x1F
+	as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, // 0x20-0x2F
+	as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, // 0x30-0x3F
+	as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, // 0x40-0x4F
+	as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, // 0x50-0x5F
+	as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, // 0x60-0x6F
+	as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, as, // 0x70-0x7F
+	//   1   2   3   4   5   6   7   8   9   A   B   C   D   E   F
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0x80-0x8F
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0x90-0x9F
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0xA0-0xAF
+	xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0xB0-0xBF
+	xx, xx, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, // 0xC0-0xCF
+	s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, s1, // 0xD0-0xDF
+	s2, s3, s3, s3, s3, s3, s3, s3, s3, s3, s3, s3, s3, s4, s3, s3, // 0xE0-0xEF
+	s5, s6, s6, s6, s7, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, // 0xF0-0xFF
+}
+
+// acceptRange gives the range of valid values for the second byte in a UTF-8
+// sequence.
+type acceptRange struct {
+	lo uint8 // lowest value for second byte.
+	hi uint8 // highest value for second byte.
+}
+
+var acceptRangesData = [32]uint8{
+	locb, hicb,
+	0xA0, hicb,
+	locb, 0x9F,
+	0x90, hicb,
+	locb, 0x8F,
+}
+
+func stdlib(d Register, n Register, ret *Basic) {
+	Comment("Non-vectorized implementation from the stdlib. Used for small inputs.")
+	mask := GP64()
+	MOVQ(U64(0x8080808080808080), mask)
+
+	first := Mem{Base: GP64(), Scale: 1}
+	LEAQ(ConstBytes("first", firstData[:]), first.Base)
+
+	acceptRanges := Mem{Base: GP64(), Scale: 2}
+	LEAQ(ConstBytes("accept_ranges", acceptRangesData[:]), acceptRanges.Base)
+
+	Comment("Fast ascii-check loop")
```

Comment:
The implementation here is likely a lot slower than the one we have in https://github.com/segmentio/asm/blob/main/ascii/valid_amd64.s, have you tried calling `ascii.Valid` as a first step in the Go code instead?

Reply:
I haven't since this was a direct translation from the stdlib code. Are you suggesting a two pass validation with ascii first and utf8 second (modified to assume everything is at least ascii)?